### PR TITLE
#380 review page and page elements

### PR DESCRIPTION
**Pull request by andreievg**
Fixes: #380 
Fixes: #349

Branched from: #386

### Changes

** PageElements** 
* Added to isReview 
* Simplified JSX for other views
* Changed prop to take PageElement (need this for thisReviewLatestResponse etc..)
* Added ReviewResponseComponent (shows one line of review)
* Added ReviewBotton (the one to edit or review the review)
** SectionWrapper**
* Simplified JSX
* Added extraSectionTitleContent prop and extraPageContent (both are methods that will render something extra, take section or page as props)
** DecisionAreaNew **
* Copied old one and simplified
** Application Page **
* getCurrentElements returns PageElement[] vs ElementState[] to (to pass to PageElements component)
** useGetFullReviewStructure **
* add reviewInfo and status to structure
** getReviewQuery **
* also returns status
** ReviewPage **
* new page to display review
* passing some components to render extra info in SectionWrapper

### Testing

* use test branch #380-testing and back end branch #386frontend-helpers
* start a review for ABC123 (with reviewer 1 or 2, using existing UI)
* navigate to http://localhost:3000/applicationNew/ABC123/reviewTest (with either of the reviewer, make sure to update test page with 1020 -> reviewer 1 or 1021 reviewer 2)
* see extra bit of information that can be use by progress in the section header, it should change as you make changes
* note, when users change review assignment should change

### CODE REVIEW

File: src/components/Application/PageElements.tsx [hunk 1]
```diff
@@ -86,13 +95,88 @@ const PageElements: React.FC<PageElementProps> = ({
...
-        })}
+        <Segment.Group>
+          {elements.map(({ element, thisReviewLatestResponse }) => (
+            <Segment key={`question_${element.id}`}>
+              <Grid columns="equal">
+                <Grid.Column floated="left">
+                  <SummaryViewWrapperNEW {...getSummaryViewProps(element)} />
+                </Grid.Column>
+                <Grid.Column floated="right" textAlign="right">
+                  <ReviewButton
```
Comment (nmadruga) [inline src/components/Application/PageElements.tsx:106]: Maybe this one should be renamed to `ApproveAllButton`?

File: src/components/Application/SectionWrapper.tsx [hunk 2]
```diff
@@ -31,30 +36,33 @@ const SectionWrapper: React.FC<SectionProps> = ({
...
-              </Grid.Column>
-            </Grid.Row>
+          <Grid columns="equal">
+            <Grid.Column floated="left">
+              <Header as="h2" content={details.title} />
+            </Grid.Column>
+            <Grid.Column floated="right" textAlign="right">
+              {extraSectionTitleContent && extraSectionTitleContent(section)}
+            </Grid.Column>
+            <Grid.Column floated="right" textAlign="right" width={1}>
```
Comment (nmadruga) [inline src/components/Application/SectionWrapper.tsx:46]:
```suggestion
            <Grid.Column>
              <Header as="h2" content={details.title} />
            </Grid.Column>
            <Grid.Column floated="right" textAlign="right">
              {extraSectionTitleContent && extraSectionTitleContent(section)}
            </Grid.Column>
            <Grid.Column floated="right" textAlign="right">
```

Similar layout problem (on mobile) found here.

**Comment by nmadruga** [comment]
> start a review for ABC123 (with reviewer 1 or 2, using existing UI)

File: src/components/Review/DecisionAreaNEW.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+              onClick={submit}
+              disabled={
+                !review.decision ||
+                (review.decision === ReviewResponseDecision.Decline && !review.comment)
+              }
+              content={strings.BUTTON_SUBMIT}
+            />
+            {review.decision === ReviewResponseDecision.Decline && !review.comment && (
+              <p style={{ color: 'red' }}>{messages.REVIEW_RESUBMIT_COMMENT}</p>
+            )}
```
Comment (nmadruga) [inline src/components/Review/DecisionAreaNEW.tsx:120]: I saw this change here. Looks good : )
Screen Shot 2021-03-06 at 5 15 54 PM

File: src/components/Application/PageElements.tsx [hunk 4]
```diff
@@ -86,13 +95,88 @@ const PageElements: React.FC<PageElementProps> = ({
...
+                    reviewResponse={thisReviewLatestResponse as ReviewResponse}
+                    summaryViewProps={getSummaryViewProps(element)}
+                  />
+                </Grid.Column>
+              </Grid>
+              <ReviewResponseComponent
+                reviewResponse={thisReviewLatestResponse as ReviewResponse}
+                summaryViewProps={getSummaryViewProps(element)}
+              />
+            </Segment>
```
Comment (nmadruga) [inline src/components/Application/PageElements.tsx:116]: The approve is looking good
Screen Shot 2021-03-06 at 5 15 43 PM

But I think the Decline need another line with the comment?
Screen Shot 2021-03-06 at 5 16 11 PM

File: src/containers/Review/ReviewPage.tsx [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+      if (!reviewResponse) return
+      updateReviewResponse({
+        variables: { id: reviewResponse.id, decision: ReviewResponseDecision.Approve },
+      })
+    })
+  }
+
+  if (responsesToReview.length === 0) return null
+
+  return <Button onClick={massApprove}>{`Approve (${responsesToReview.length})`}</Button>
```
Comment (nmadruga) [inline src/containers/Review/ReviewPage.tsx:99]: Please use localised strigns

**Review by nmadruga — CHANGES_REQUESTED** [review]
Great work! It's good to see this page in place : D

And I made a few layout suggestions, also there is another one that I noticed in the PageElements for Review that seems weird (not sure if the same things is happening on the ApplicationPage). INFORMATION elements the content should display using full row lenght:
Screen Shot 2021-03-06 at 5 15 05 PM

File: src/components/Application/PageElements.tsx (hunk 4 above)
Comment (andreievg) [reply src/components/Application/PageElements.tsx:116]: 3c4f0d27981fe01c5c961b5d2d51c44b2c1b5e5c

Sorry not making a huge effort on these styles atm

File: src/containers/Review/ReviewPage.tsx (hunk 5 above)
Comment (andreievg) [reply src/containers/Review/ReviewPage.tsx:99]: 3c4f0d27981fe01c5c961b5d2d51c44b2c1b5e5c

File: src/components/Application/PageElements.tsx [hunk 6]
```diff
@@ -1,14 +1,22 @@
-import React from 'react'
+import React, { useState } from 'react'
 import { Link } from 'react-router-dom'
-import { ElementStateNEW, ResponsesByCode, SectionAndPage } from '../../utils/types'
+import { ElementStateNEW, PageElement, ResponsesByCode, SectionAndPage } from '../../utils/types'
 import ApplicationViewWrapper from '../../formElementPlugins/ApplicationViewWrapperNEW'
 import SummaryViewWrapperNEW from '../../formElementPlugins/SummaryViewWrapperNEW'
-import { Form, Grid, Segment, Button } from 'semantic-ui-react'
+import { Form, Grid, Segment, Button, Card, Icon, Label } from 'semantic-ui-react'
```
[the remark this answers is not in the document]
Comment (andreievg) [reply src/components/Application/PageElements.tsx]: cdf97857d1a065498e47035d0bb20cb280fc630a

File: src/components/Application/PageElements.tsx [hunk 7]
```diff
@@ -1,14 +1,22 @@
...
 import strings from '../../utils/constants'
-import { TemplateElementCategory } from '../../utils/generated/graphql'
+import {
+  ReviewResponse,
+  ReviewResponseDecision,
+  TemplateElementCategory,
+  useUpdateReviewResponseMutation,
+} from '../../utils/generated/graphql'
```
[the remark this answers is not in the document]
Comment (andreievg) [reply src/components/Application/PageElements.tsx]: cdf97857d1a065498e47035d0bb20cb280fc630a

File: src/components/Application/SectionWrapper.tsx (hunk 2 above)
Comment (andreievg) [reply src/components/Application/SectionWrapper.tsx:46]: i feel like the progress bar is shifted too much to the left this way. hmm, was really hoping to sort out styling towards the end (i mean if you check out the rest of the app in iPhone 5 mode...), the reason I even changed them in this pr was to simplify jsx. How do you feel about leaving this one for fixing layout stage stage ?

File: src/components/Application/PageElements.tsx [hunk 8]
```diff
@@ -46,37 +54,38 @@ const PageElements: React.FC<PageElementProps> = ({
...
+            return (
+              <Segment key={`question_${element.code}`}>
+                <Grid columns="equal">
+                  <Grid.Column floated="left">
+                    <SummaryViewWrapperNEW {...getSummaryViewProps(element)} />
                   </Grid.Column>
-                  <Grid.Column width={3}>
-                    {element.category === TemplateElementCategory.Question && canEdit && (
+                  {element.category === TemplateElementCategory.Question && canEdit && (
+                    <Grid.Column width={3} floated="right">
```
[the remark this answers is not in the document]
Comment (andreievg) [reply src/components/Application/PageElements.tsx]: thanks for noticing: cdf97857d1a065498e47035d0bb20cb280fc630a

File: src/components/Application/PageElements.tsx (hunk 1 above)
Comment (andreievg) [reply src/components/Application/PageElements.tsx]: 5221d0610db840c66350011d38ea137b82178e3e

**Comment by andreievg** [comment]
@nmadruga i think i've dealt with all of the requests apart from section title layout one (left a comment), thanks for review

File: src/utils/helpers/structure/generateReviewProgress.ts [hunk 9]
```diff
@@ -73,15 +73,14 @@ const getSums = (elements: PageNEW[]) => {
...
-return elements.reduce(
+  return elements.reduce(
     (sum, { reviewProgress }) => ({
       totalReviewable: sum.totalReviewable + (reviewProgress?.totalReviewable || 0),
       doneConform: sum.doneConform + (reviewProgress?.doneConform || 0),
       doneNoneConform: sum.doneNoneConform + (reviewProgress?.doneNoneConform || 0),
     }),
     initial
   )
-  )
```
Comment (CarlosNZ) [inline src/utils/helpers/structure/generateReviewProgress.ts:84]: Weird that you can remove one bracket and not break something. 🤷‍♂️

**Review by CarlosNZ — CHANGES_REQUESTED** [review]
Cool, not completely clear why you suggested testing with the test branch and `/ReviewTest` as opposed to the proper URL, but I'm going to assume that the "proper" branch is the same?

Still getting one compile error:
```
ERROR in [at-loader] ./src/components/Review/ReviewPageWrapperTest.tsx:13:56 
    TS2322: Type '{ status: ReviewStatus.Draft; }' is not assignable to type 'Maybe<Review>'.
  Property 'nodeId' is missing in type '{ status: ReviewStatus.Draft; }' but required in type 'Node'.
```

Other than that, I think it's good to go.

File: src/utils/helpers/structure/generateReviewProgress.ts (hunk 9 above)
Comment (andreievg) [reply src/utils/helpers/structure/generateReviewProgress.ts]: I think that was broken in the branched this was branched from (and subsequently fixed, and just now merged to master), and after update to master i can't see that file in list of change)

**Comment by andreievg** [comment]
> Cool, not completely clear why you suggested testing with the test branch and /ReviewTest as opposed to the proper URL, but I'm going to assume that the "proper" branch is the same?

this was before review_assignment info was generated, so needed to fake that being passed through, there is a page wrapper that's also required (testing with test branch just passed through a review_assignment_id, hard coded)

As for compiling errors (on latest master server, after database_init, and after regenerating, had a different compilation error), now fixed ab70069f10bdd55f6e1512ef00a52943e078c5dc

**Review by CarlosNZ — APPROVED** [review]

**Comment by andreievg** [comment]
@CarlosNZ thanks for review

**Pull request merged**